File: app/src/main/java/com/example/rentalwheels/PlaceorderActivity.java
package com.example.rentalwheels;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.app.AlertDialog;
import android.app.DatePickerDialog;
import android.app.TimePickerDialog;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.AutoCompleteTextView;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.TextView;
import android.widget.TimePicker;
import android.widget.Toast;

import com.example.rentalwheels.models.AppUser;
import com.example.rentalwheels.models.DBHelper;
import com.example.rentalwheels.models.DBModel;
import com.example.rentalwheels.models.UserOrders;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.android.material.timepicker.MaterialTimePicker;
import com.google.android.material.timepicker.TimeFormat;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.FirebaseDatabase;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;

import timber.log.Timber;

public class PlaceorderActivity extends AppCompatActivity {

    AutoCompleteTextView autoCompleteTextView;
    private EditText doc_details;
    String docdetails;
    int Hours = 0;
    int Days = 0;
    int Audi = 0;
    int LandCruiser = 0;
    int BMW = 0;
    TextView topQty, lowerQty, timeview, topPrice, lowerPrice;
    public RadioGroup radioService;
    public RadioButton radiobtn;
    private DatePickerDialog datePickerDialog;
    private MaterialTimePicker picker;
    private Calendar calendar;
    Button checkout;
    int t2hour, t2minute;
    int month, date, year, hours, minutes;
    String time;
    int cart_total = 0;
    ProgressBar progressbar;
    DBHelper dbHelper;
    List<DBModel> dbModelList = new ArrayList<>();
    Button dateButton;





    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_placeorder);
        overridePendingTransition(R.anim.abc_fade_in,R.anim.abc_fade_out);

        dbHelper = new DBHelper(this);
        topQty = findViewById(R.id.topQty);
        lowerQty = findViewById(R.id.lowerQty);
        autoCompleteTextView = findViewById(R.id.autoComplete);
        radioService = findViewById(R.id.radioService);
        dateButton = findViewById(R.id.datebutton);
        timeview = findViewById(R.id.timeview);
        checkout = findViewById(R.id.checkout);
        topPrice = findViewById(R.id.topPrice);
        lowerPrice = findViewById(R.id.lowerPrice);
        dbModelList = dbHelper.getDataFromDbForHistory();
        doc_details = findViewById(R.id.docdetails);


        String [] option = {"Passport", "National ID" , "VISA"};
        ArrayAdapter arrayAdapter = new ArrayAdapter(this, R.layout.list_item, option);
        autoCompleteTextView.setText(arrayAdapter.getItem(0).toString(), false);
        autoCompleteTextView.setAdapter(arrayAdapter);

        //checkout
        checkout.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                int Hours = Integer.parseInt(topPrice.getText().toString());
                int Days = Integer.parseInt(lowerPrice.getText().toString());
                String docdetails = doc_details.getText().toString().trim();
                String day = date + "/" + month + "/" + year;
                if (docdetails.isEmpty() || docdetails.length() <6){
                    Toast.makeText(PlaceorderActivity.this, "Oops not yet, Kindly recheck if your details", Toast.LENGTH_LONG).show();
                    return;
                }
                if (hours == 0) {
                    hours = 12;
                    time = hours + ":" + minutes;

                } else {
                    time = hours + ":" + minutes;

                }
                Timber.e("onClick: " + Hours + "\n" + Days + "\n"  + day + "\n" + time);
                cart_total = Hours + Days ;
                if (Hours == 0 && Days == 0 ) {
                    Toast.makeText(PlaceorderActivity.this, "cart can't be empty", Toast.LENGTH_SHORT).show();
                    return;
                }
               else {
                    dbHelper.insert(Hours, Days,
                            Audi, BMW, LandCruiser, cart_total, day, time);
                    dbHelper.insertTemp(Hours, Days,
                            Audi, BMW, LandCruiser, cart_total, day, time);
                    SharedPreferences sp = getApplicationContext().getSharedPreferences("RentalWheels", Context.MODE_PRIVATE);
                    dbHelper.insertUserOrder(sp.getString("email", ""), Hours, Days,
                            Audi, BMW, LandCruiser, cart_total, day, time);

                }
                startActivity(new Intent(PlaceorderActivity.this, CheckoutActivity.class));

            }
        });

        //timepicker
        timeview.setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View v){
                showTimePicker();
            }
        });


        //datepicker
        initDatePicker();
        dateButton.setText(getTodayDate());
    }

    private void showTimePicker() {
        picker = new MaterialTimePicker.Builder()
                .setTimeFormat(TimeFormat.CLOCK_12H)
                .setHour(12)
                .setMinute(0)
                .setTitleText("Set Reminder Time")
                .build();

        picker.show(getSupportFragmentManager(), "RentalWheels");

        picker.addOnPositiveButtonClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                if (picker.getHour() > 12){
                    timeview.setText(
                            String.format("%02d",(picker.getHour()-12))+ ":" +String.format("%02d", picker.getMinute()) + "PM"
                    );
                }else {
                    timeview.setText(picker.getHour() + ":" + picker.getMinute() + "AM");
                }

                calendar = Calendar.getInstance();
                calendar.set(Calendar.HOUR_OF_DAY, picker.getHour());
                calendar.set(Calendar.MINUTE, picker.getMinute());
                calendar.set(Calendar.SECOND, 0);
                calendar.set(Calendar.MILLISECOND, 0);
            }
        });
    }

    private String getTodayDate(){
        Calendar cal = Calendar.getInstance();
        int year = cal.get(Calendar.YEAR);
        int month = cal.get(Calendar.MONTH);
        month = month + 1;
        int day = cal.get(Calendar.DAY_OF_MONTH);
        return makeDateString(day, month, year);
    }
    //date
    private void initDatePicker(){
        DatePickerDialog.OnDateSetListener dateSetListener = new DatePickerDialog.OnDateSetListener() {
            @Override
            public void onDateSet(DatePicker datePicker, int year, int month, int day) {
                month = month + 1;
                String date = makeDateString(day, month, year);
            }
        };

        Calendar cal = Calendar.getInstance();
        int year = cal.get(Calendar.YEAR);
        int month = cal.get(Calendar.MONTH);
        int day = cal.get(Calendar.DAY_OF_MONTH);

        int style = AlertDialog.THEME_HOLO_LIGHT;
        datePickerDialog = new DatePickerDialog(this, style, dateSetListener, year, month, day);
//        datePickerDialog.getDatePicker().setMinDate(System.currentTimeMillis());

    }

    private String makeDateString(int day, int month, int year){
        return getMonthFormat(month) + " " + day + " " + year;
    }

    private String getMonthFormat(int month){
        if (month == 1)
            return "JAN";
        if (month == 2)
            return "FEB";
        if (month == 3)
            return "MAR";
        if (month == 4)
            return "APR";
        if (month == 5)
            return "MAY";
        if (month == 6)
            return "JUN";
        if (month == 7)
            return "JUL";
        if (month == 8)
            return "AUG";
        if (month == 9)
            return "SEP";
        if (month == 10)
            return "OCT";
        if (month == 11)
            return "NOV";
        if (month == 12)
            return "DEC";

        //default should never happen
        return "JAN";
    }

    public void openDatePicker(View view){
        datePickerDialog.show();
    }

    public void clickradioButton(View view) {
        boolean checked = ((RadioButton) view).isChecked();
        // Check which radio button was clicked
        switch (view.getId()) {
            case R.id.Audi:
                if (checked) {
                    displayHourPrice(Hours * 100);
                    displayDailyCharge(Days * 2000);
                    Audi = 1;
                    BMW = 0;
                    LandCruiser = 0;
                }
                break;
            case R.id.BMW:
                if (checked) {
                    displayHourPrice(Hours * 150);
                    displayDailyCharge(Days * 2500);
                    Audi = 0;
                    BMW = 1;
                    LandCruiser = 0;
                }
                break;
            case R.id.LandCruiser:
                if (checked) {

                    displayHourPrice(Hours * 200);
                    displayDailyCharge(Days * 3000);
                    Audi = 0;
                    BMW = 0;
                    LandCruiser = 1;
                }
                break;
        }
    }

    //adding
    public void topInc(View view) {
        if (Hours == 11) {
            Hours = 11;
            Toast.makeText(getApplicationContext(), "Can't be more than 11", Toast.LENGTH_SHORT).show();
        } else {
            Hours = Hours + 1;
            displayHour(Hours);

        }
        updatePrice("Hours", Hours);
    }

    public void topDec(View view) {
        if (Hours == 0) {
            Hours = 0;
            Toast.makeText(getApplicationContext(), "Can't be less than 0", Toast.LENGTH_SHORT).show();
        } else {
            Hours = Hours - 1;
            displayHour(Hours);
        }
        updatePrice("Hours", Hours);
    }

    public void lowerInc(View view) {
        if (Days == 10) {
            Days = 10;
            Toast.makeText(getApplicationContext(), "Can't be more than 10", Toast.LENGTH_SHORT).show();
        } else {
            Days = Days + 1;
            displayDaily(Days);
        }
        updatePrice("Days", Days);
    }

    public void lowerDec(View view) {
        if (Days == 0) {
            Days = 0;
            Toast.makeText(getApplicationContext(), "Can't be less than 0", Toast.LENGTH_SHORT).show();
        } else {
            Days = Days - 1;
            displayDaily(Days);
        }
        updatePrice("Daily", Days);
    }

    private void displayHour(int number) {

        topQty.setText("" + number);
    }

    private void displayDaily(int number) {
        lowerQty.setText("" + number);
    }

    public void updatePrice(String orderType, int count) {
        int selectedId = radioService.getCheckedRadioButtonId();
        // find the radiobutton by returned id
        radiobtn = (RadioButton) findViewById(selectedId);
        if (radiobtn.getText().equals("Audi")) {
            if (orderType == "Hour") {
                displayHourPrice(count * 5);
            } else if (orderType == "Daily") {
                displayDailyCharge(count * 10);
            }
        } else if (radiobtn.getText().equals("BMW")) {
            if (orderType == "Hour") {
                displayHourPrice(count * 3);
            } else if (orderType == "Daily") {
                displayDailyCharge(count * 3);
            }
        } else if (radiobtn.getText().equals("LandCruiser")) {
            if (orderType == "Hour") {
                displayHourPrice(count * 8);
            } else if (orderType == "Daily") {
                displayDailyCharge(count * 13);
            }
        }
    }


    private void displayHourPrice(int number) {
        topPrice.setText("" + number);
    }

    private void displayDailyCharge(int number) {
        lowerPrice.setText("" + number);
    }

}

File: app/src/main/java/com/msdc/rentalwheels/modules/PlaceorderActivity.kt
package com.msdc.rentalwheels.modules

import android.app.AlertDialog
import android.app.DatePickerDialog
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import com.msdc.rentalwheels.R
import com.msdc.rentalwheels.models.DBHelper
import com.msdc.rentalwheels.models.DBModel
import com.google.android.material.timepicker.MaterialTimePicker
import com.google.android.material.timepicker.TimeFormat
import timber.log.Timber
import java.util.*

class PlaceorderActivity : AppCompatActivity() {

    private lateinit var autoCompleteTextView: AutoCompleteTextView
    private lateinit var docDetails: EditText
    private lateinit var topQty: TextView
    private lateinit var lowerQty: TextView
    private lateinit var timeView: TextView
    private lateinit var topPrice: TextView
    private lateinit var lowerPrice: TextView
    private lateinit var radioService: RadioGroup
    private lateinit var datePickerDialog: DatePickerDialog
    private lateinit var picker: MaterialTimePicker
    private lateinit var calendar: Calendar
    private lateinit var checkout: Button
    private lateinit var progressBar: ProgressBar
    private lateinit var dbHelper: DBHelper
    private lateinit var dateButton: Button

    private var hours = 0
    private var days = 0
    private var audi = 0
    private var landCruiser = 0
    private var bmw = 0
    private var cartTotal = 0
    private var time: String? = null
    private var dbModelList: List<DBModel> = ArrayList()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_placeorder)
        overridePendingTransition(R.anim.abc_fade_in, R.anim.abc_fade_out)

        dbHelper = DBHelper(this)
        topQty = findViewById(R.id.topQty)
        lowerQty = findViewById(R.id.lowerQty)
        autoCompleteTextView = findViewById(R.id.autoComplete)
        radioService = findViewById(R.id.radioService)
        dateButton = findViewById(R.id.datebutton)
        timeView = findViewById(R.id.timeview)
        checkout = findViewById(R.id.checkout)
        topPrice = findViewById(R.id.topPrice)
        lowerPrice = findViewById(R.id.lowerPrice)
        docDetails = findViewById(R.id.docdetails)

        val options = arrayOf("Passport", "National ID", "VISA")
        val arrayAdapter = ArrayAdapter(this, R.layout.list_item, options)
        autoCompleteTextView.setText(arrayAdapter.getItem(0).toString(), false)
        autoCompleteTextView.setAdapter(arrayAdapter)

        checkout.setOnClickListener {
            val hours = topPrice.text.toString().toInt()
            val days = lowerPrice.text.toString().toInt()
            val docDetails = docDetails.text.toString().trim()
            val cal = Calendar.getInstance()
            val day = "${cal.get(Calendar.DAY_OF_MONTH)}/${cal.get(Calendar.MONTH) + 1}/${cal.get(Calendar.YEAR)}"

            if (docDetails.isEmpty() || docDetails.length < 6) {
                Toast.makeText(this, "Oops not yet, Kindly recheck if your details", Toast.LENGTH_LONG).show()
                return@setOnClickListener
            }

            time = if (hours == 0) {
                "12:${cal.get(Calendar.MINUTE)}"
            } else {
                "$hours:${cal.get(Calendar.MINUTE)}"
            }

            val currentTime = time ?: ""
            Timber.e("onClick: $hours\n$days\n$day\n$currentTime")
            cartTotal = hours + days

            if (hours == 0 && days == 0) {
                Toast.makeText(this, "cart can't be empty", Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            } else {
                dbHelper.insert(hours, days, audi, bmw, landCruiser, cartTotal, day, currentTime)
                dbHelper.insertTemp(hours, days, audi, bmw, landCruiser, cartTotal, day, currentTime)
                val sp = getSharedPreferences("RentalWheels", Context.MODE_PRIVATE)
                dbHelper.insertUserOrder(sp.getString("email", "") ?: "", hours, days, audi, bmw, landCruiser, cartTotal, day, currentTime)
            }

            startActivity(Intent(this, CheckoutActivity::class.java))
        }

        timeView.setOnClickListener { showTimePicker() }

        initDatePicker()
        dateButton.text = getTodayDate()
    }

    private fun showTimePicker() {
        picker = MaterialTimePicker.Builder()
            .setTimeFormat(TimeFormat.CLOCK_12H)
            .setHour(12)
            .setMinute(0)
            .setTitleText("Set Reminder Time")
            .build()

        picker.show(supportFragmentManager, "RentalWheels")

        picker.addOnPositiveButtonClickListener {
            timeView.text = if (picker.hour > 12) {
                String.format("%02d", picker.hour - 12) + ":" + String.format("%02d", picker.minute) + "PM"
            } else {
                picker.hour.toString() + ":" + picker.minute + "AM"
            }

            calendar = Calendar.getInstance().apply {
                set(Calendar.HOUR_OF_DAY, picker.hour)
                set(Calendar.MINUTE, picker.minute)
                set(Calendar.SECOND, 0)
                set(Calendar.MILLISECOND, 0)
            }
        }
    }

    private fun getTodayDate(): String {
        val cal = Calendar.getInstance()
        val year = cal.get(Calendar.YEAR)
        val month = cal.get(Calendar.MONTH) + 1
        val day = cal.get(Calendar.DAY_OF_MONTH)
        return makeDateString(day, month, year)
    }

    private fun initDatePicker() {
        val dateSetListener = DatePickerDialog.OnDateSetListener { _, year, month, day ->
            val date = makeDateString(day, month + 1, year)
        }

        val cal = Calendar.getInstance()
        val year = cal.get(Calendar.YEAR)
        val month = cal.get(Calendar.MONTH)
        val day = cal.get(Calendar.DAY_OF_MONTH)

        val style = AlertDialog.THEME_HOLO_LIGHT
        datePickerDialog = DatePickerDialog(this, style, dateSetListener, year, month, day)
    }

    private fun makeDateString(day: Int, month: Int, year: Int): String {
        return "${getMonthFormat(month)} $day $year"
    }

    private fun getMonthFormat(month: Int): String {
        return when (month) {
            1 -> "JAN"
            2 -> "FEB"
            3 -> "MAR"
            4 -> "APR"
            5 -> "MAY"
            6 -> "JUN"
            7 -> "JUL"
            8 -> "AUG"
            9 -> "SEP"
            10 -> "OCT"
            11 -> "NOV"
            12 -> "DEC"
            else -> "JAN"
        }
    }

    fun openDatePicker(view: View) {
        datePickerDialog.show()
    }

    fun clickRadioButton(view: View) {
        if (view is RadioButton) {
            when (view.id) {
                R.id.Audi -> if (view.isChecked) {
                    displayHourPrice(hours * 100)
                    displayDailyCharge(days * 2000)
                    audi = 1
                    bmw = 0
                    landCruiser = 0
                }
                R.id.BMW -> if (view.isChecked) {
                    displayHourPrice(hours * 150)
                    displayDailyCharge(days * 2500)
                    audi = 0
                    bmw = 1
                    landCruiser = 0
                }
                R.id.LandCruiser -> if (view.isChecked) {
                    displayHourPrice(hours * 200)
                    displayDailyCharge(days * 3000)
                    audi = 0
                    bmw = 0
                    landCruiser = 1
                }
            }
        }
    }

    fun topInc(view: View) {
        if (hours == 11) {
            Toast.makeText(this, "Can't be more than 11", Toast.LENGTH_SHORT).show()
        } else {
            hours++
            displayHour(hours)
        }
        updatePrice("Hours", hours)
    }

    fun topDec(view: View) {
        if (hours == 0) {
            Toast.makeText(this, "Can't be less than 0", Toast.LENGTH_SHORT).show()
        } else {
            hours--
            displayHour(hours)
        }
        updatePrice("Hours", hours)
    }

    fun lowerInc(view: View) {
        if (days == 10) {
            Toast.makeText(this, "Can't be more than 10", Toast.LENGTH_SHORT).show()
        } else {
            days++
            displayDaily(days)
        }
        updatePrice("Days", days)
    }

    fun lowerDec(view: View) {
        if (days == 0) {
            Toast.makeText(this, "Can't be less than 0", Toast.LENGTH_SHORT).show()
        } else {
            days--
            displayDaily(days)
        }
        updatePrice("Days", days)
    }

    private fun displayHour(number: Int) {
        topQty.text = number.toString()
    }

    private fun displayDaily(number: Int) {
        lowerQty.text = number.toString()
    }

    private fun updatePrice(orderType: String, count: Int) {
        val selectedId = radioService.checkedRadioButtonId
        val radioButton = findViewById<RadioButton>(selectedId)
        when (radioButton.text) {
            "Audi" -> {
                if (orderType == "Hours") {
                    displayHourPrice(count * 5)
                } else if (orderType == "Days") {
                    displayDailyCharge(count * 10)
                }
            }
            "BMW" -> {
                if (orderType == "Hours") {
                    displayHourPrice(count * 3)
                } else if (orderType == "Days") {
                    displayDailyCharge(count * 3)
                }
            }
            "LandCruiser" -> {
                if (orderType == "Hours") {
                    displayHourPrice(count * 8)
                } else if (orderType == "Days") {
                    displayDailyCharge(count * 13)
                }
            }
        }
    }

    private fun displayHourPrice(number: Int) {
        topPrice.text = number.toString()
    }

    private fun displayDailyCharge(number: Int) {
        lowerPrice.text = number.toString()
    }
}